app/include/drivers/behavior.h

```diff
@@ -23,6 +23,64 @@
  * (Internal use only.)
  */
 
+#if IS_ENABLED(CONFIG_ZMK_BEHAVIOR_METADATA)
+
+enum behavior_parameter_standard_domain {
+    BEHAVIOR_PARAMETER_STANDARD_DOMAIN_NULL = 0,
+    BEHAVIOR_PARAMETER_STANDARD_DOMAIN_HID_USAGE = 1,
+    BEHAVIOR_PARAMETER_STANDARD_DOMAIN_LAYER_INDEX = 2,
+    BEHAVIOR_PARAMETER_STANDARD_DOMAIN_HSV = 3,
+};
+
+struct behavior_parameter_value_metadata {
+    char *friendly_name;
+    uint8_t position;
+
+    union {
+        uint32_t value;
+        struct {
+            uint16_t min;
+            uint16_t max;
+        } range;
+
+        enum behavior_parameter_standard_domain standard;
+    };
+
+    enum {
+        BEHAVIOR_PARAMETER_VALUE_METADATA_TYPE_VALUE = 0,
+        BEHAVIOR_PARAMETER_VALUE_METADATA_TYPE_RANGE = 1,
+        BEHAVIOR_PARAMETER_VALUE_METADATA_TYPE_STANDARD = 2,
+    } type;
+};
+
+struct behavior_parameter_metadata_custom_set {
+    size_t values_len;
+    const struct behavior_parameter_value_metadata *values;
+};
+
+struct behavior_parameter_metadata_custom {
+    size_t sets_len;
+    struct behavior_parameter_metadata_custom_set sets[];
+};
+
+struct behavior_parameter_metadata {
+    union {
+        struct {
+            enum behavior_parameter_standard_domain param1;
+            enum behavior_parameter_standard_domain param2;
+        } standard;
```

Comment:
If I'm understanding this correctly, this is effectively just a shortcut for 

```c
{
  .type = BEHAVIOR_PARAMETER_METADATA_CUSTOM,
  .custom = {
    .sets_len = 1,
    .sets = {{
      .values_len = 2,
      .values = {{
        .position = 0,
        .type = BEHAVIOR_PARAMETER_VALUE_METADATA_TYPE_STANDARD,
        .standard = <param1>,
      }, {
        .position = 1,
        .type = BEHAVIOR_PARAMETER_VALUE_METADATA_TYPE_STANDARD,
        .standard = <param2>,
      }},
    }},
  },
}
```

right? If so, that simplifies defining the data, but it adds a special case to the implementation that has to read this data. You also still have to use the more complex form for something like a MIDI note behavior that uses `param1 = range[0, 127]` (note) and `param2 = range[0, 127]` (velocity), since there's no way to represent a range type here.

Is there any way we could avoid the redundant ways to represent the same data? Maybe have some predefined constants and/or macros to let you build a metadata set that only has one item without all the boilerplate? Alternatively, could we support all possible value types in this inline form of writing the parameter data?

Reply:
So... this is a very valid question..... After adding the necessary complexity for custom params.... do we really need both ways of doing this? I think I'm almost convinced we don't. Let me sleep on this, but if I can't come up with a compelling reason, I will refactor to "one single way of doing this" with some naming tweaks to make this saner.

Reply:
Ok, pushed a major refactor here to simplify. Thoughts?